feat(lua): use callable table as iterator in vim.iter (#23957)
neovim/neovim: runtime/doc/lua.txt

• Non-list tables pass both the key and value of each element
• Function iterators pass all of the values returned by their respective
  function

> • Tables with with a metatable implementing __call is treated as function
>   iterator

Reviewer: ```suggestion
• Tables with a metatable implementing `__call` are treated as function
  iterators
```
